ee/server/service/vpp.go

```diff
@@ -25,9 +27,116 @@ func (svc *Service) getVPPToken(ctx context.Context) (string, error) {
+	}
+
+	exp, err := time.Parse("2006-01-02T15:04:05Z0700", vppTokenRaw.ExpDate)
+	if err != nil {
+		return "", ctxerr.Wrap(ctx, err, "parsing vpp token expiration date")
+	}
+
+	if time.Now().After(exp) {
+		return "", ctxerr.Errorf(ctx, "vpp token expired on %s", exp.String())
+	}
+
 	return base64.StdEncoding.EncodeToString([]byte(vppTokenData.Token)), nil
 }
 
+func (svc *Service) BatchAssociateVPPApps(ctx context.Context, teamName string, payloads []fleet.VPPBatchPayload, dryRun bool) error {
+	if teamName == "" {
+		svc.authz.SkipAuthorization(ctx) // so that the error message is not replaced by "forbidden"
+		return ctxerr.Wrap(ctx, fleet.NewInvalidArgumentError("team_name", "must not be empty"))
+	}
+
+	if err := svc.authz.Authorize(ctx, &fleet.Team{}, fleet.ActionRead); err != nil {
+		return err
+	}
+
+	team, err := svc.ds.TeamByName(ctx, teamName)
+	if err != nil {
+		// If this is a dry run, the team may not have been created yet
+		if dryRun && fleet.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+
+	if team == nil {
+		return ctxerr.Wrap(ctx, fleet.NewInvalidArgumentError("team_name", "team does not exist"))
+	}
+
+	if err := svc.authz.Authorize(ctx, &fleet.SoftwareInstaller{TeamID: &team.ID}, fleet.ActionWrite); err != nil {
+		return ctxerr.Wrap(ctx, err, "validating authorization")
+	}
+
+	token, err := svc.getVPPToken(ctx)
+	if err != nil {
+		return fleet.NewUserMessageError(ctxerr.Wrap(ctx, err, "could not retrieve vpp token"), http.StatusUnprocessableEntity)
+	}
+
+	if len(payloads) == 0 {
+		return fleet.NewUserMessageError(ctxerr.Errorf(ctx, "request must contain at least one assset"), http.StatusBadRequest)
+	}
+
+	var adamIDs []string
+
+	for _, payload := range payloads {
+		adamIDs = append(adamIDs, payload.AppStoreID)
+	}
+
+	var missingAssets []string
+	var validAssets []vpp.Asset
+
+	assets, err := vpp.GetAssets(token, nil)
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "unable to retrieve assets")
+	}
+
+	for _, adamID := range adamIDs {
+		var found bool
+		for _, asset := range assets {
+			if asset.AdamID == adamID {
+				found = true
+				validAssets = append(validAssets, asset)
+				continue
+			}
+		}
+		if !found {
+			missingAssets = append(missingAssets, adamID)
+		}
+	}
+
+	if len(missingAssets) != 0 {
+		reqErr := ctxerr.Errorf(ctx, "requested app not available on vpp account: %s", strings.Join(missingAssets, ","))
+		return fleet.NewUserMessageError(reqErr, http.StatusUnprocessableEntity)
+	}
+
+	// we cheked if the team is null earlier
+	serials, err := svc.ds.GetTeamAppleSerialNumbers(ctx, team.ID)
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "cannot get team serials for association")
+	}
+
+	// There is nothing to assiciate, and calling the apple API with
+	// no serial numbers returns an error
+	if len(serials) == 0 {
+		return nil
+	}
+
+	if !dryRun {
+		if err := vpp.AssociateAssets(token, &vpp.AssociateAssetsRequest{
```

Comment:
I'm a bit confused, maybe @roperzh or @jahzielv can chime in as they've more context on the story than I do but my understanding is that the batch-assign VPP apps was to add an app to the team so that it can be installed on hosts, and to do that it would insert (or delete) into `vpp_apps_teams` table. It should only accept teams that exist in `vpp_apps` table (but maybe it makes sense to call `GetAssets` to refresh that list at that point in time?).

The batch endpoint's role is to replace the whole list of vpp apps for a given team with whatever is received (which could be empty, in which case it would delete all apps for that team).

Associating assets with serials, AIUI, is only done when attempting to install an app on a specific host, and is done just for that specific host? See https://www.figma.com/design/QsO3Q03QGVbiSknGpLqgAB/%2318867-Install-Apple-App-Store-apps-on-macOS?node-id=367-7396&t=JoJmxOLeHI0SbceD-0. Otherwise we run into issues that associating in this batch endpoint associates with ALL current hosts (which might consume all licenses, if licenses are marked as "consumed" when associating?), and future changes in hosts' team membership is not taken into account.